File: wakeonlan/src/main/java/com/github/shingyx/wakeonlan/SendPacketTask.java
package com.github.shingyx.wakeonlan;

import android.os.AsyncTask;

public class SendPacketTask extends AsyncTask<Void, Void, Void> {
    private MagicPacketProcessor magicPacketProcessor;
    private String macAddress;
    private Callback callback;

    public SendPacketTask(MagicPacketProcessor magicPacketProcessor, String macAddress, Callback callback) {
        this.magicPacketProcessor = magicPacketProcessor;
        this.macAddress = macAddress;
        this.callback = callback;
    }

    public SendPacketTask(MagicPacketProcessor magicPacketProcessor, Callback callback) {
        this(magicPacketProcessor, magicPacketProcessor.getSavedMacAddress(), callback);
    }

    @Override
    protected Void doInBackground(Void... voids) {
        String error = null;
        try {
            magicPacketProcessor.send(macAddress);
        } catch (Exception e) {
            error = e.getMessage();
        }
        callback.function(error);
        return null;
    }
}

File: wakeonlan/src/main/java/com/github/shingyx/wakeonlan/SendPacketTask.kt
package com.github.shingyx.wakeonlan

import android.os.AsyncTask

class SendPacketTask(
    private val magicPacketProcessor: MagicPacketProcessor,
    private val macAddress: String,
    private val callback: (error: String?) -> Unit
) : AsyncTask<Unit, Unit, String?>() {
    constructor(
        magicPacketProcessor: MagicPacketProcessor,
        callback: (error: String?) -> Unit
    ) : this(magicPacketProcessor, magicPacketProcessor.getSavedMacAddress(), callback)

    override fun doInBackground(vararg args: Unit): String? {
        return try {
            magicPacketProcessor.send(macAddress)
            null
        } catch (e: Exception) {
            e.message
        }
    }

    override fun onPostExecute(result: String?) {
        callback(result)
    }
}
